nucliadb_node/src/bin/writer.rs

```diff
@@ -147,25 +147,32 @@ pub async fn watch_node_update(
     info!("Start node update task");
 
     while let Some(event) = receiver.recv().await {
-        match event {
-            NodeWriterEvent::ShardCreation => {
-                metadata.shard_count = metadata.shard_count.saturating_add(1);
-                info!("Update node state: shard_count = {}", metadata.shard_count);
-                node.update_state(SHARD_COUNT_KEY, metadata.shard_count)
-                    .await;
+        let (load_score, shard_count) = match event {
+            NodeWriterEvent::ShardCreation(id) => {
+                metadata.new_empty_shard(id);
+
+                (None, Some(metadata.shard_count()))
             }
-            NodeWriterEvent::ShardDeletion => {
-                metadata.shard_count = metadata.shard_count.saturating_sub(1);
-                info!("Update node state: shard_count = {}", metadata.shard_count);
-                node.update_state(SHARD_COUNT_KEY, metadata.shard_count)
-                    .await;
+            NodeWriterEvent::ShardDeletion(id) => {
+                metadata.delete_shard(id);
+
+                (Some(metadata.load_score()), Some(metadata.shard_count()))
             }
-            NodeWriterEvent::ParagraphCount(paragraph_count) => {
-                metadata.paragraph_count = paragraph_count;
-                info!("Update node state: load_score = {paragraph_count}");
-                node.update_state(LOAD_SCORE_KEY, paragraph_count as f32)
-                    .await;
+            NodeWriterEvent::ParagraphCount(id, paragraph_count) => {
+                metadata.update_shard(id, paragraph_count);
+
+                (Some(metadata.load_score()), None)
             }
+        };
+
+        if let Some(load_score) = load_score {
+            info!("Update node state: load_score = {load_score}");
+            node.update_state(LOAD_SCORE_KEY, load_score).await;
+        }
+
+        if let Some(shard_count) = shard_count {
+            info!("Update node state: shard_count = {shard_count}");
+            node.update_state(SHARD_COUNT_KEY, shard_count).await;
         }
```

Comment:
Maybe the following approach would simplify your loop:
```rust
async fn update_node_state<T: std::fmt::Display>(node: &NodeHandle, key: Key<T>, score: T) {
    info!("Update node state: {} = {score}", key.as_str());
    node.update_state(key, score).await;
}

pub async fn watch_node_update(
    node: NodeHandle,
    mut receiver: UnboundedReceiver<NodeWriterEvent>,
    mut metadata: NodeWriterMetadata,
) {
    info!("Start node update task");

    while let Some(event) = receiver.recv().await {
        match event {
            NodeWriterEvent::ShardCreation(id) => {
                metadata.new_empty_shard(id);
                update_node_state(&node, SHARD_COUNT_KEY, metadata.shard_count()).await;
            }
            NodeWriterEvent::ShardDeletion(id) => {
                metadata.delete_shard(id);
                update_node_state(&node, LOAD_SCORE_KEY, metadata.load_score()).await;
                update_node_state(&node, SHARD_COUNT_KEY, metadata.shard_count()).await;
            }
            NodeWriterEvent::ParagraphCount(id, paragraph_count) => {
                metadata.update_shard(id, paragraph_count);
                update_node_state(&node, LOAD_SCORE_KEY, metadata.load_score()).await;
            }
        }
    }

    info!("Node update task stopped");
}
```

Reply:
Nice suggestion :cat:

Reply:
Done at 807bf855cfe52417caf208cb76ee1b30f2c7598c